sdk/storage/Azure.Storage.Queues/src/QueueClientExtensions.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Azure.Storage.Queues.Specialized
+{
+    /// <summary>
+    /// Provides advanced extensions for queue service clients.
+    /// </summary>
+    public static class QueueClientExtensions
+    {
+        /// <summary>
+        /// Creates a new instance of the <see cref="QueueClient"/> class, maintaining all the same
+        /// internals but specifying new <see cref="ClientSideEncryptionOptions"/>.
+        /// </summary>
+        /// <param name="client">Client to base off of.</param>
+        /// <param name="clientSideEncryptionOptions">New encryption options. Setting this to <code>default</code> will clear client-side encryption.</param>
+        /// <returns>New instance with provided options and same internals otherwise.</returns>
+        public static QueueClient WithClientSideEncryptionOptions(this QueueClient client, ClientSideEncryptionOptions clientSideEncryptionOptions)
+            => new QueueClient(
+                client.Uri,
+                client.Pipeline,
+                client.Version,
+                client.ClientDiagnostics,
+                clientSideEncryptionOptions,
+                client.OnClientSideDecryptionFailure);
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="QueueClient"/> class, maintaining all the same
+        /// internals but specifying new <see cref="ClientSideEncryptionOptions"/>.
+        /// </summary>
+        /// <param name="client">Client to base off of.</param>
+        /// <param name="listener">Listener for when decryption of a single message fails.</param>
+        /// <returns>New instance with provided options and same internals otherwise.</returns>
+        public static QueueClient WithClientSideEncryptionFailureListener(this QueueClient client, IClientSideDecryptionFailureListener listener)
```

Comment:
If it's possible to have extension method taking optional parameters then I'd consider consolidating these extension into `WithClientSideEncryption(this client, options, listener = default)`.

Reply:
I originally had this and decided against it. There is no easy way to distinguish between leaving one of the objects as is versus clearing it. I'd rather have the customer set each separately than force them to maintain a reference to one when they only want to change the other. Since the whole point of this method is to provide a lightweight mechanism to swap the key, I am not worried about someone who actually wants to change both having to call into 2 separate lightweight methods.

Reply:
gotcha. so we want to prevent client.WithClientEcryption(options, litenere).WithClientEncryption(options) from zeroing listener. Makes sense.